apps/src/block_utils.js

```diff
@@ -380,6 +380,35 @@ exports.cleanBlocks = function (blocksDom) {
   });
 };
 
+/**
+ * Adds any functions from functionsXml to blocksXml. If a function with the
+ * same name is already present in blocksXml, it won't be added again.
+ */
+exports.appendNewFunctions = function (blocksXml, functionsXml) {
+  const startBlocksDom = xml.parseElement(blocksXml);
+  const sharedFunctionsDom = xml.parseElement(functionsXml);
+  const functions = [...sharedFunctionsDom.ownerDocument.firstElementChild.children];
+  for (let func of functions) {
+    const name = func.ownerDocument.evaluate(
+      'title[@name="NAME"]/text()', func, null, XPathResult.STRING_TYPE,
+    ).stringValue;
+    const type = func.ownerDocument.evaluate(
+      '@type', func, null, XPathResult.STRING_TYPE,
+    ).stringValue;
+    const alreadyPresent = startBlocksDom.ownerDocument.evaluate(
+      `//block[@type="${type}"]/title[@name="NAME"][text()="${name}"]`,
+      startBlocksDom, null, XPathResult.UNORDERED_NODE_SNAPSHOT_TYPE,
+    ).snapshotLength > 0;
```

Comment:
I sort of like the simplicity of manipulating the XML directly, without a Blockly dependency.  That said, would it be better to have this logic live in Blockly, or use Blockly's methods for procedure blocks to manipulate the block space?

Willing to be convinced otherwise!

Reply:
I wanted to keep this change small since I might change the implementation later. The "pretend that shared behaviors are just part of the start blocks" approach seems to be the simplest, and doesn't touch much.